# Msc_project_supplementary
#### Bioinformatics final project codes and other supplementary materials
#
#
### Figures:
###   1. COVID_vs_gnomad_vep_impact.pdf
####    - Evolutionary age distribution of different impact variants for each COVID-19 GWAS study and its control 
####    - High, low, moderate and modifier denote variant impact with different severity. The first three are coding-region related variants,             while modifiers are non-coding region related variants. 
####    - The difference between COVID-19 and the controls in different impact groups was tested using a non-parametric Wilcoxon–Mann–Whitney             test (****: p < 0.0001, ***: p < 0.001, **: p < 0.01, NS: not significant).

###   2. all_species_allele_beta_plot.pdf
####    - Empirical cumulative distribution function plots of beta values in each GWAS for every species
####    - The dashed line at y = 0.5 is where the median can be found for each beta distribution.
#
#
### Statistics:
###   1. wilcoxon_beta folder
####     - Pairwise wilcoxon comparison for ancestral beta values for each GWAS in each species 




